integration/backup_and_restore_test.go

```diff
@@ -0,0 +1,212 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+//go:build integration
+
+package integration
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/caarlos0/env/v6"
+	"github.com/eclipse-kanto/file-backup/client"
+	fut "github.com/eclipse-kanto/file-upload/integration"
+	"github.com/eclipse-kanto/file-upload/uploaders"
+	"github.com/eclipse-kanto/kanto/integration/util"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/net/websocket"
+)
+
+type backupAndRestoreParams struct {
+	Options map[string]string `json:"Options"`
+}
+
+func newBackupAndRestoreParams(dir string, downloadURL string) (*backupAndRestoreParams, error) {
+	absBackupDirPath, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, err
+	}
+	return &backupAndRestoreParams{
+		Options: map[string]string{
+			client.BackupDirProperty: absBackupDirPath,
+			uploaders.URLProp:        downloadURL,
+		},
+	}, nil
+}
+
+func (suite *fileBackupSuite) SetupSuite() {
+	suite.Setup(suite.T())
+
+	opts := env.Options{RequiredIfNoDef: false}
+	suite.backupAndRestoreCfg = backupAndRestoreConfig{}
+	require.NoError(suite.T(), env.Parse(&suite.backupAndRestoreCfg, opts), "failed to process backup and restore environment variables")
+	suite.T().Logf("backup and restore test configuration - %v", suite.backupAndRestoreCfg)
+	suite.AssertEmptyDir(suite.backupAndRestoreCfg.BackupDir)
+	suite.AssertEmptyDir(suite.backupAndRestoreCfg.RestoreDir)
+
+	suite.ThingURL = util.GetThingURL(suite.Cfg.DigitalTwinAPIAddress, suite.ThingCfg.DeviceID)
+	suite.FeatureURL = util.GetFeatureURL(suite.ThingURL, featureID)
+}
+
+func (suite *fileBackupSuite) TearDownSuite() {
+	suite.TearDown()
+}
+
+func TestHTTPFileBackup(t *testing.T) {
+	suite.Run(t, new(httpFileBackupSuite))
+}
+
+func TestAzureFileBackup(t *testing.T) {
+	suite.Run(t, new(azureFileBackupSuite))
+}
+
+func TestAWSFileBackup(t *testing.T) {
+	suite.Run(t, new(awsFileBackupSuite))
+}
+
+func (suite *httpFileBackupSuite) TestFileBackup() {
+	suite.SetupStorageProvider(fut.GenericStorageProvider)
+	defer suite.TearDownStorageProvider()
+	suite.testBackupAndRestore()
+}
+
+func (suite *azureFileBackupSuite) TestFileBackup() {
+	suite.SetupStorageProvider(fut.AzureStorageProvider)
+	defer suite.TearDownStorageProvider()
+	suite.testBackupAndRestore()
+}
+
+func (suite *awsFileBackupSuite) TestFileBackup() {
+	suite.SetupStorageProvider(fut.AWSStorageProvider)
+	defer suite.TearDownStorageProvider()
+	suite.testBackupAndRestore()
+}
+
+func (suite *fileBackupSuite) testBackupAndRestore() {
+	backupDir := suite.backupAndRestoreCfg.BackupDir
+	restoreDir := suite.backupAndRestoreCfg.RestoreDir
+	defer suite.RemoveFilesSilently(backupDir)
+	defer suite.RemoveFilesSilently(restoreDir)
```

Comment:
Why "silently"? Does not this means that the tests can leak directories and files? What would be the impact if this happens?

Reply:
It means, that even if a file deletion fails, the test continues, removing the other files(without failing).

Reply:
I get that, I was wondering if it makes sense to aggregate the error(s) from the deletions and fail the test with one aggregated error instead of just logging it. Do you think it will be beneficial or it would be simply an overkill?

Reply:
Seems like an overkill, IMO.